includes/Core/Modules/Module.php

```diff
@@ -433,17 +433,25 @@ public function is_connected() {
 	 * @since 1.0.0
 	 *
 	 * @param Data_Request $data Data request object.
-	 *
 	 * @return mixed Data on success, or WP_Error on failure.
+	 *
+	 * phpcs:disable Squiz.Commenting.FunctionCommentThrowTag.Missing
 	 */
 	final protected function execute_data_request( Data_Request $data ) {
-		$client     = $this->get_client();
-		$orig_defer = $client->shouldDefer();
-		$client->setDefer( true );
+		$datapoint_services = $this->get_datapoint_services();
+
+		// We only need to initialize the client if this datapoint relies on a service.
+		if ( ! empty( $datapoint_services[ $data->datapoint ] ) ) {
+			$client     = $this->get_client();
+			$orig_defer = $client->shouldDefer();
+			$client->setDefer( true );
+		}
 
 		$request = $this->create_data_request( $data );
 
-		$client->setDefer( $orig_defer );
+		if ( isset( $client ) ) {
+			$client->setDefer( $orig_defer );
+		}
```

Comment:
This pattern of backing up the set defer and then creating/executing a request is a pattern that is used quite a bit in the codebase. It generally adds noise to what's going on and is something I wish was a bit cleaner every time I see it. This would be great to see this extracted to a helper method.

Something like this perhaps?
```php
public static function with_defer( Client $client, $defer ) {
    $orig_defer = $client->shouldDefer();
    $client->setDefer( $defer );

    // Return a function to restore the original refer value.
    return function () use ( $client, $orig_defer ) {
        $client->setDefer( $orig_defer );
    };
}
```

In this case, it would be simplified like so:

```php
if ( ! empty( $datapoint_services[ $data->datapoint ] ) ) {
    $restore_defer = static::with_defer( $this->get_client(), true );
}

$request = $this->create_data_request( $data );

if ( isset( $restore_defer ) ) {
    $restore_defer();
}
```

I'm not sure which class this method would best belong to, but it's purely functional so that's more of a decision of aesthetic/expressiveness as it is.